toolshed/index.ts

```diff
@@ -22,23 +23,93 @@ const initializeStorage = async () => {
 
 export type AppType = typeof app;
 
-// Initialize and start the server
-const startServer = async () => {
-  // Initialize signer before starting server
-  await initializeStorage();
+// Graceful shutdown with timeout
+const ac = new AbortController();
+let isShuttingDown = false;
+const SHUTDOWN_TIMEOUT = 10000; // 10 seconds
+
+const handleShutdown = async () => {
+  if (isShuttingDown) {
+    console.log("Shutdown already in progress...");
+    return;
+  }
+
+  isShuttingDown = true;
+  console.log("Shutdown signal received, closing server...");
+
+  // Create a timeout promise
+  const timeoutPromise = new Promise((_, reject) => {
+    setTimeout(() => reject(new Error("Shutdown timed out")), SHUTDOWN_TIMEOUT);
+  });
+
+  try {
+    // Race between graceful shutdown and timeout
+    await Promise.race([
+      (async () => {
+        // Remove signal listeners to prevent multiple shutdown attempts
+        Deno.removeSignalListener("SIGINT", handleShutdown);
+        Deno.removeSignalListener("SIGTERM", handleShutdown);
 
-  console.log(`Server is running on port http://localhost:${port}`);
+        ac.abort();
+
+        console.log("Closing memory system...");
+        const result = await memory.close();
+        if (result.error) {
+          console.error("Error closing memory:", result.error);
+        } else {
+          console.log("Memory system closed successfully");
+        }
+      })(),
+      timeoutPromise,
+    ]);
+  } catch (err) {
+    console.error("Error during shutdown:", err);
+  }
+
+  console.log("Shutdown complete");
+  Deno.exit(0);
+};
+
+// Start server with the abort controller
+async function startServer() {
+  console.log(`Server is starting on port http://${env.HOST}:${env.PORT}`);
+  await initializeStorage();
 
   Sentry.init({
     dsn: env.SENTRY_DSN,
     tracesSampleRate: 1.0,
+    environment: env.ENV || "development",
   });
 
-  Deno.serve({ port }, app.fetch);
-};
+  const serverOptions = {
+    hostname: env.HOST,
+    port: env.PORT,
+    signal: ac.signal,
+    onError: (error: unknown) => {
+      console.error("Server error:", error);
+      Sentry.captureException(error);
+      return new Response("Internal Server Error", { status: 500 });
+    },
+    onListen: ({ port, hostname }: { port: number; hostname: string }) => {
+      console.log(`Server running on http://${hostname}:${port}`);
+    },
+  };
+
+  try {
+    Deno.serve(serverOptions, app.fetch);
+  } catch (err) {
+    if (err instanceof Deno.errors.AddrInUse) {
+      console.error(`Port ${env.PORT} is already in use`);
+      Deno.exit(1);
+    }
+    console.error("Failed to start server:", err);
+    Deno.exit(1);
+  }
+}
+
+Deno.addSignalListener("SIGINT", handleShutdown);
+Deno.addSignalListener("SIGTERM", handleShutdown);
 
-// Start server and handle errors
-startServer().catch((error) => {
-  console.error("Failed to start server:", error);
-  Deno.exit(1);
-});
+if (import.meta.main) {
```

Comment:
Can this even be used as a library?

Reply:
nope .. but I sometimes use it in deno dev mode...

Reply:
err, the repl